app/controllers/invoices_controller.rb

```diff
@@ -0,0 +1,112 @@
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+
+class InvoicesController < ApplicationController
+
+  PER_PAGE = 30
+
+  before_action :require_login, except: %i[show]
+  before_action :unlocked_user, except: %i[show]
+  before_action :select_bootstrap41
+  before_action :show_user_navbar, except: %i[show]
+
+  def new
+    load_invoiceable
+    @invoice = Invoice.new
+  end
+
+  def create
+    load_invoiceable
+    @invoice = @invoiceable.invoices.build(invoice_create_params)
+    @invoice.slug = SecureRandom.urlsafe_base64(32)
+    @invoice.user_id = current_user.id
+    @invoice.status = 'open'
+    if @invoice.save
+      flash[:success] = "Invoice Generated"
+      redirect_to polymorphic_path([@invoiceable.adopter])
+    else
+      flash.now[:error] = "Invoice could not be saved"
+      render 'new'
+    end
+  end
+
+  def show
+    @invoice = Invoice.friendly.find(params[:id])
+  end
+
+  def edit
+  end
+
+  def update
+    @invoice = Invoice.friendly.find(params[:id])
+    @invoice.card_token =  stripe_params["stripeToken"]
+
+    begin
+      @invoice.process_payment(@invoice.invoiceable.adopter.email)
+    rescue Stripe::CardError => e
+      flash[:error] = e.message
+    else
+      @invoice.donation = stripe_params["invoice"]["donation"]
+      @invoice.paid_method = 'Stripe'
+      @invoice.paid_at = Time.now
+      @invoice.status = 'paid'
+      @invoice.save
+      @donation = Donation.new
+      @donation.name = @invoice.invoiceable.adopter.name
+      @donation.email = @invoice.invoiceable.adopter.email
+      @donation.amount = @invoice.donation
+      @donation.frequency = 'Once'
+      @donation.comment = 'Adoption Fee Roundup'
+      @donation.save
+      flash[:success] = 'Payment Processed Successfully '
+      InvoiceMailer.invoice_paid(@invoice.id).deliver_later
+    ensure
+      render :show
+    end
+  end
+
+  def index
+    @invoices = Invoice.order(created_at: :desc).paginate(page: params[:page], per_page: format_for_page)
+  end
+
+  private
+
+  def format_for_page
+    return PER_PAGE unless request.format.xls?
+
+    Invoice.count
+  end
+
+  def invoice_create_params
+    params.require(:invoice).permit(:amount, :description)
+  end
+
+  def stripe_params
+    params.permit :stripeToken,
+                  :utf8,
+                  :authenticity_token,
+                  :_method,
+                  :id,
+                  invoice:
+                  [
+                    :donation
+                  ]
+  end
+
+  def load_invoiceable
+    resource, id = request.path.split('/')[1, 2]
```

Comment:
how come you aren't doing this with routes?  Also I'd check resource is a value that is allowed here, allowing user input to select the class doing a find makes me nervous, also anyway you can scope it to the current user's invoices?

Reply:
An `Invoice` is polymorphic, and can belong to an `Adoption` or a `CatAdoption`, so this was needed to find the owning object.    As stolen from here: "The Ryan Bates Method" (RailsCast Dude)  https://codecabulary.launchacademy.com/learn-rails/polymorphic-associations

This could be waaay out of date way to handle it, but I hadn't had much luck finding better examples.